rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMRecoverer.java

```diff
@@ -0,0 +1,294 @@
+import org.ow2.proactive.resourcemanager.core.properties.PAResourceManagerProperties;
+import org.ow2.proactive.resourcemanager.db.NodeSourceData;
+import org.ow2.proactive.resourcemanager.db.RMDBManager;
+import org.ow2.proactive.resourcemanager.db.RMNodeData;
+import org.ow2.proactive.resourcemanager.nodesource.NodeSource;
+import org.ow2.proactive.resourcemanager.rmnode.RMNode;
+
+
+/**
+ * This classe handles the recovery of node sources and of nodes when the RM is restarted.
+ *
+ * @author ActiveEon Team
+ * @since 26/06/17
+ */
+public class RMRecoverer {
+
+    private static final Logger logger = Logger.getLogger(RMRecoverer.class);
+
+    public static final String START_TO_RECOVER_NODES = "Start to recover nodes";
+
+    public static final String END_OF_NODES_RECOVERY = "Total number of nodes recovered: ";
+
+    private RMCore resourceManager;
+
+    private RMDBManager dbManager;
+
+    public RMRecoverer(RMCore resourceManager, RMDBManager dbManager) {
+        this.resourceManager = resourceManager;
+        this.dbManager = dbManager;
+    }
+
+    protected void initiateRecoveryIfRequired() {
+        if (nodesRecoveryEnabled()) {
+            logger.info("Starting Nodes Recovery");
+        } else {
+            logger.info("Nodes Recovery is disabled. Removing all nodes from database");
+            dbManager.removeAllNodes();
+        }
+        recoverNodeSourcesAndNodes();
+    }
+
+    /**
+     * @return whether there are nodes in database for the given node source
+     *
+     * In case of a recovery, we need to check whether there are nodes in
+     * database for this node source, otherwise, we will do a redeployment
+     * from scratch. The reason is that when the RM shuts down correctly,
+     * it removes all its nodes. Thus if we restart and recover the RM
+     * afterwards there will be no nodes in the database.
+     */
+    public boolean existNodesToRecover(String nodeSourceName, boolean nodesRecoverable) {
+        boolean recoverNodes = false;
+        if (nodesRecoveryEnabled() && nodesRecoverable) {
+            Collection<RMNodeData> nodesData = dbManager.getNodesByNodeSource(nodeSourceName);
+            if (nodesData.isEmpty()) {
+                logger.info("No node found in database for node source: " + nodeSourceName);
+            } else {
+                recoverNodes = true;
+            }
+        }
+        return recoverNodes;
+    }
+
+    private void recoverNodeSourcesAndNodes() {
+        Collection<NodeSourceData> nodeSources = dbManager.getNodeSources();
+        logPersistedNodeSourceInfo(nodeSources);
+
+        for (NodeSourceData nodeSourceData : nodeSources) {
+            String nodeSourceDataName = nodeSourceData.getName();
+            try {
+                logger.info("Recovering node source " + nodeSourceDataName);
+                resourceManager.createNodeSource(nodeSourceData, nodeSourceData.getNodesRecoverable());
+            } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)